src/Plutarch/SafeMoney.hs

```diff
@@ -178,10 +187,99 @@ pdiscreteValue' (Tagged (AssetClass (cs, tn))) =
             PDiscrete t <- pmatchC p
             pure $ psingletonValue # pconstant cs # pconstant tn # (pextract # t)
 
-{-
-pdiscreteValue' (Tagged (AssetClass (cs, tn))) = phoistAcyclic $
-  plam $ \p ->
-    psingletonValue
-      # pconstant cs
-      # pconstant tn
-      # Tagged.puntag p -}
+-- | @since 1.0.1
+type PRateDecimal = PFixedDecimal 1000000
+
+-- | @since 1.0.1
+newtype PExchangeRate (from :: k) (to :: k') (s :: S)
+    = PExchangeRate (Term s (PTagged '(from, to) PRateDecimal))
+
+deriveGeneric ''PExchangeRate
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PlutusType (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PIsData (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        PEq (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving via
+    (DerivePNewtype (PExchangeRate from to) (PTagged '(from, to) PRateDecimal))
+    instance
+        POrd (PExchangeRate from to)
+
+-- | @since 1.0.1
+deriving anyclass instance PShow (PExchangeRate from to)
+
+-- | @since 1.0.1
+instance PTryFrom PData (PAsData (PExchangeRate from to)) where
+    type
+        PTryFromExcess PData (PAsData (PExchangeRate from to)) =
+            PTryFromExcess PData (PAsData PRateDecimal)
+    ptryFrom' d k = ptryFrom' @_ @(PAsData PRateDecimal) d $ k . first punsafeCoerce
+
+{- | Make @PExchangeRation@ with given @PIntegers. Two Integers are nominator and denominator.
+
+ @since 1.0.1
+-}
+pexchangeRatio ::
+    forall k.
+    forall (to :: k) (from :: k) (s :: S).
+    Term s (PInteger :--> PInteger :--> PMaybe (PExchangeRate from to))
+pexchangeRatio = phoistAcyclic $ plam $ \n d -> pexchangeRatio' n d
+
+{- | Same as @pexchangeRation@ but in Haskell level.
+
+ @since 1.0.1
+-}
+pexchangeRatio' ::
+    forall k.
+    forall (to :: k) (from :: k) (s :: S).
+    Term s PInteger ->
+    Term s PInteger ->
+    Term s (PMaybe (PExchangeRate from to))
+pexchangeRatio' n d =
+    pif
+        (d #== 0)
+        (pcon PNothing)
+        ( pcon . PJust $
+            pcon . PExchangeRate $
+                ppure #$ (fromPInteger # n) `divide` (fromPInteger # d)
+        )
```

Comment:
Do we need both of these? I feel that if we have a Plutarch-level function, we can get a `Term`-level one using `#` just fine.